### PR TITLE
Extend metadata with descriptors of CF Ref handling

### PR DESCRIPTION
**Pull request by t0yv0**
CloudFormation Ref intrinsic behavior varies from resource to resource. This change is an attempt to schematize behavior and expose is in the provider metadata. One intended use case is `pulumi/pulumi-cdk` CF emulator.

This change implies the `ref-db.json` becomes the source of truth for this metadata and we maintain it as well as possible.

About 1/2 of CF resources are currently covered in the file, including 81 of the top used resources (https://github.com/pulumi/pulumi-cdk/issues/237). Data cleaning and data label attachment helper code is included but is not intended to be used in production but only to maintain `ref-db.json`.

### CODE REVIEW

File: meta/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+Reference data used by the `pulumi-aws-native` provider.
+
+## ref-db.json
+
+Metadata describing the behavior of the [CloudFormation Ref Intrinsic](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-ref.html).
+
+What Ref returns varies from resource to resource. The only available information on this is free-form text in CloudFormation documentation and anything that can be extracted experimentally. To facilitate building tooling such as `pulumi/pulumi-cdk` that emulates CF behavior, `ref-db.json` collects schematized information describing the behavior of `Ref`.
+
+`ref-db.json` can be edited directly to correct errors or add entries for resources not yet covered.
```
Comment (corymhall) [inline meta/README.md:11]: Can you add an example here on what it looks like to update this?

**Review by corymhall — APPROVED** [review]
Looks good! Most of my comments are just around documentation. This may be something that should live in pulumi-cdk, but we can always move it later.

File: provider/tools/ref-parser/cat.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+type categorizationRule struct {
+	Category Category
+	Pattern  *regexp.Regexp
+}
+
+var categorizationRules = []categorizationRule{
+	{
+		Category: RefReturnsArn,
+		Pattern:  regexp.MustCompile(`^The Amazon Resource Name ..ARN.. of the certificate authority.?.?$`),
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply provider/tools/ref-parser/cat.go:36]: This is not supposed to be scalable, this is throw-away data cleaning code. I'll add a quick README. Once we complete curating a clean data set in ref-db.json we can delete this.

File: meta/README.md (hunk 1 above)
Comment (t0yv0) [reply meta/README.md:11]: You send a PR and edit it? I'm not sure I understand. It just has to conform to the schema.

File: provider/tools/ref-parser/autolabel.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+		if _, seen := db.Resources[r]; seen {
+			continue
+		}
+
+		sch, err := findSchema(schemaAbsPath, r)
+		if err != nil {
+			continue
+		}
+
+		switch {
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply provider/tools/ref-parser/autolabel.go:33]: It's a bit ad-hoc but I've added a quick comment.

**Pull request merged**